packages/network/src/logic/InstructionSender.ts

```diff
@@ -0,0 +1,108 @@
+ * when either of these conditions is satisfied: 
+ * - the topology stabilizes: no new instructions has been formed for the stream in 
+ *   X milliseconds
+ * - the buffer times out: we have buffered an instruction for Y milliseconds
+ * 
+ * When an instruction is added to a stream buffer, it may overwrite an existing
+ * instruction in the buffer if the both instructions share the same nodeId. In that 
+ * situation we expect that the previous instruction is no longer valid (it has a lower
+ * counterValue) and can be ignored.
+ */
+
+const DEFAULT_TOPOLOGY_STABILIZATION: TopologyStabilizationOptions = {
+    debounceWait: 100,
+    maxWait: 2000
+}
+
+const logger = new Logger(module)
+
+export interface Instruction {
+    nodeId: NodeId,
+    streamKey: StreamKey,
+    newNeighbors: NodeId[],
+    counterValue: number
+}
+
+class StreamInstructionBuffer {
+    private instructions: Map<NodeId,Instruction> = new Map()
+    private debouncedOnReady: () => void
+
+    constructor(topologyStabilization: TopologyStabilizationOptions, onReady: () => void) {
+        this.debouncedOnReady = _.debounce(onReady, topologyStabilization.debounceWait, {
+            maxWait: topologyStabilization.maxWait
+        })
+    }
+
+    addInstruction(instruction: Instruction) {
+        // may overwrite an earlier instruction for the same node
+        this.instructions.set(instruction.nodeId, instruction)
+        this.debouncedOnReady()
+    }
+
+    getInstructions(): IterableIterator<Instruction> {
+        return this.instructions.values()
+    }
+}
+
+export class InstructionSender {
+
+    private readonly streamBuffers: Map<StreamKey,StreamInstructionBuffer> = new Map()
+    private readonly topologyStabilization: TopologyStabilizationOptions
+    private readonly trackerServer: TrackerServer
+    private readonly metrics: Metrics
+
+    constructor(topologyStabilization: TopologyStabilizationOptions|undefined, trackerServer: TrackerServer, metrics: Metrics) {
+        this.topologyStabilization = topologyStabilization ?? DEFAULT_TOPOLOGY_STABILIZATION
+        this.trackerServer = trackerServer
+        this.metrics = metrics
+    }
+
+    addInstruction(instruction: Instruction) {
+        this.getOrCreateBuffer(instruction.streamKey).addInstruction(instruction)
+    }
+
+    getOrCreateBuffer(streamKey: StreamKey) {
+        const existingBuffer = this.streamBuffers.get(streamKey)
+        if (existingBuffer !== undefined) {
+            return existingBuffer
+        } else {
+            const newBuffer = new StreamInstructionBuffer(this.topologyStabilization, () => {
+                this.streamBuffers.delete(streamKey)
+                this.sendInstructions(newBuffer)
+            })
+            this.streamBuffers.set(streamKey, newBuffer)
+            return newBuffer
+        }
+    }
+
+    private async sendInstructions(buffer: StreamInstructionBuffer) {
+        for (const instruction of buffer!.getInstructions()) {
```

Comment:
what happens, or should happen if it's still executing another `sendInstructions`? I guess the counterValue disambiguates and gives an ordering, but perhaps it would be good to cancel any pending instructions that are superseded?

Reply:
If the node receives an old instruction (a `counterValue` which is smaller than the most recent one), it will just ignore it. Also, if we form a new instruction for the same stream+node while we buffer the previous instruction, it is automatically overwritten in the buffer. 

Maybe it is quite unlikely that we'd be sending a buffer at the same time when a new instruction is formed?
- the deboucing reduces the amount of instructions significantly
- the buffer has just stabilized or timeouted -> if it stabilized, but we still form a new instruction right after the stabilization, we should maybe increase the `debounceWait` configuration option
- the sending of the buffer is a relatively fast operation